### PR TITLE
Redux whiteboard

### PR DESCRIPTION
**Pull request by taylorpwhipp**
### Description

This PR migrates the redux-whiteboard example to typescript

#### How to test it

Just like all the other examples: simple to test by running `npm run dev` and checking `http://localhost:3000`

#### Related issue(s)

- Add the related issue(s) here:
  - https://github.com/liveblocks/[REPOSITORY]/issues/1073

### CODE REVIEW

File: examples/nextjs-redux-whiteboard/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+<details><summary>Read more</summary>
+
+<p></p>
+
+Alternatively, you can set up your project manually:
+
+- Install all dependencies with `npm install`
+- Create an account on [liveblocks.io](https://liveblocks.io/dashboard)
+- Copy your **public** key from the [dashboard](https://liveblocks.io/dashboard/apikeys)
+- - Create a`.env.local` file and add your **public** key as the
```
Comment (CTNicholas) [inline examples/nextjs-redux-whiteboard/README.md:55]:
```suggestion
- Create a`.env.local` file and add your **public** key as the
```

File: examples/nextjs-redux-whiteboard/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  </a>
+  <a href="https://codesandbox.io/s/github/liveblocks/liveblocks/tree/main/examples/nextjs-redux-whiteboard">
+    <img src="https://img.shields.io/badge/open%20in%20codesandbox-message?style=flat&logo=codesandbox&color=333&logoColor=fff" alt="Open in CodeSandbox" />
+  </a>
+  <img src="https://img.shields.io/badge/react-message?style=flat&logo=react&color=0bd&logoColor=fff" alt="React" />
+  <img src="https://img.shields.io/badge/redux-message?style=flat&logo=redux&color=74b&logoColor=fff" alt="Redux" />
+</p>
+
+This example shows how to build a collaborative whiteboard with
+[Liveblocks](https://liveblocks.io), [React](https://reactjs.org/) and
```
Comment (CTNicholas) [inline examples/nextjs-redux-whiteboard/README.md:24]:
```suggestion
[Liveblocks](https://liveblocks.io), [Next.js](https://nextjs.org/) and
```

File: examples/nextjs-redux-whiteboard/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+<details><summary>Read more</summary>
+
+<p></p>
+
+To both deploy on [Vercel](https://vercel.com), and run the example locally, use
+the following command:
+
+```bash
+npx create-liveblocks-app@latest --example nextjs-dashboard --vercel
```
Comment (CTNicholas) [inline examples/nextjs-redux-whiteboard/README.md:69]:
```suggestion
npx create-liveblocks-app@latest --example nextjs-redux-whiteboard --vercel
```

File: examples/nextjs-redux-whiteboard/package.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "@liveblocks/nextjs-redux-whiteboard",
+  "description": "This example shows how to build a collaborative whiteboard with Liveblocks, Next.js React and Redux.",
```
Comment (CTNicholas) [inline examples/nextjs-redux-whiteboard/package.json:3]:
```suggestion
  "description": "This example shows how to build a collaborative whiteboard with Liveblocks, Next.js and Redux.",
```

**Review by CTNicholas — COMMENTED** [review]
Noticed a couple of minor things when quickly reading over the files. Will download and run the project later!

**Comment by CTNicholas** [comment]
Could you rename this to `redux-whiteboard`? Then we can match the others (e.g. `zustand-flowchart` is a Next.js project, but uses Zustand). Changes should be needed in package.json, folder name, room id, README info etc. Thank you @taylorpwhipp!

**Comment by CTNicholas** [comment]
Just opened this and realised it isn't ready for review yet, as it's all still `.js`, whoops 👍 let me know when to look.

File: examples/nextjs-redux-whiteboard/pages/index.tsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  Shape,
+  State,
+  User
+} from "../src/store";
+import styles from "./app.module.css";
+
+let roomId = "nextjs-redux-whiteboard";
+
+export default function MyApp() {
+  useOverrideRoomId("roomId");
```
Comment (CTNicholas) [inline examples/nextjs-redux-whiteboard/pages/index.tsx:21]:
```suggestion
let roomId = "redux-whiteboard";
useOverrideRoomId();

export default function MyApp() {
```
First change to overrideRoomId

File: examples/nextjs-redux-whiteboard/pages/index.tsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+ * You can ignore it completely if you run the example locally.
+ */
+function useOverrideRoomId(roomId: string) {
+  const { query } = useRouter();
+  const overrideRoomId = useMemo(() => {
+    return query?.roomId ? `${roomId}-${query.roomId}` : roomId;
+  }, [query, roomId]);
+
+  return overrideRoomId;
+}
```
Comment (CTNicholas) [inline examples/nextjs-redux-whiteboard/pages/index.tsx:148]:
```suggestion
function overrideRoomId() {
  if (typeof window === "undefined") {
    return;
  }

  const query = new URLSearchParams(window.location?.search);
  const roomIdSuffix = query.get("roomId");

  if (roomIdSuffix) {
    roomId = `${roomId}-${roomIdSuffix}`;
  }
}
```
Second change to overrideRoomId.

Next.js runs this code on the client _and_ on the server, but we only want this to run on the client, so we're checking if window exists, and then modifying the roomId from the URL.

**Review by CTNicholas — COMMENTED** [review]
Okay, we need to make a change to `overrideRoomId` in this one! So, explanation:

We use this so that on [these pages](https://liveblocks.io/examples/collaborative-whiteboard/nextjs/preview?roomId=x-UJ5fU5-XYkUd0ULCwTq), we can override the room id from the URL. We need the `roomId` to be ready immediately for `actions.enterRoom(roomId)`, and also we need the id in another component, so we can switch the `overrideRoomId` function for a different one.

You can check it's working by adding `?roomId=another-room-id` to your URL and seeing if you're in a new room.

**Comment by CTNicholas** [comment]
Types look good! I took a look at the Redux docs, and they recommend a pattern we can implement to improve it more, [see here](https://redux.js.org/usage/usage-with-typescript#define-typed-hooks). I DMed you an explanation, check that out for more info!

**Comment by taylorpwhipp** [comment]
> Could you rename this to `redux-whiteboard`? Then we can match the others (e.g. `zustand-flowchart` is a Next.js project, but uses Zustand). Changes should be needed in package.json, folder name, room id, README info etc. **Note that my suggestions above may need to be changed too.** Thank you @taylorpwhipp! :)

Yep!

**Review by CTNicholas — APPROVED** [review]
Pull main and good to go!

**Pull request merged**